src/solver/opamSolver.ml

```diff
@@ -772,19 +772,27 @@ let dump_universe universe oc =
 
 let filter_solution filter t =
   let t = OpamCudf.ActionGraph.copy t in
-  let rec rm iter_deps v =
-    if OpamCudf.ActionGraph.mem_vertex t v then (
-      iter_deps (rm iter_deps) t v;
+  let rec rm ?(check=fun _ -> true) iter_deps v =
```

Comment:
it is not clear from the implementation what the new `check` parameter checks for. Could this be renamed something like `is_not_already_set_for_removal`? (if i understand the intent correctly)